tesseract/tesseract_planning/tesseract_motion_planners/include/tesseract_motion_planners/descartes/impl/descartes_tesseract_kinematics.hpp

```diff
@@ -0,0 +1,170 @@
+#include <tesseract_motion_planners/descartes/descartes_tesseract_kinematics.h>
+#include <Eigen/Eigen>
+#include <console_bridge/console.h>
+
+namespace tesseract_motion_planners
+{
+template <typename FloatType>
+bool DescartesTesseractKinematics<FloatType>::ik(const Eigen::Transform<FloatType, 3, Eigen::Isometry>& p,
+                                                 std::vector<FloatType>& solution_set) const
+{
+  return ik(p, is_valid_fn_, redundant_sol_fn_, solution_set);
+}
+
+template <typename FloatType>
+bool DescartesTesseractKinematics<FloatType>::ik(
+    const Eigen::Transform<FloatType, 3, Eigen::Isometry>& p,
+    const descartes_light::IsValidFn<FloatType>& is_valid_fn,
+    const descartes_light::GetRedundantSolutionsFn<FloatType>& redundant_sol_fn,
+    std::vector<FloatType>& solution_set) const
+{
+  unsigned int dof = tesseract_ik_->numJoints();
+
+  // Convert to appropriate Eigen types
+  Eigen::Isometry3d p_double;
+  p_double = p.template cast<double>();
+  Eigen::VectorXd solution_eigen;
+
+  // Solve IK
+  if (!tesseract_ik_->calcInvKin(solution_eigen, p_double, ik_seed_))
+    return false;
+
+  // Convert back to a float array
+  Eigen::Matrix<FloatType, Eigen::Dynamic, 1> solution_float_type;
+  solution_float_type = solution_eigen.template cast<FloatType>();
+  FloatType* sol = solution_float_type.data();
+
+  // Apply is_valid_fn and redundant_sol_fn
+
+  // Levi - Do we need harmonizeTowardZero(sol) here?
```

Comment:
I would only do this in the case if a redundant solution function was not provided and it failed the is valid function.

Reply:
Now if it fails the is_valid_fn it calls harmonizeTowardsZero and then retries is_valid_fn once. Note that if the number is > 3pi this will not necessarily get it to < pi at the moment. See https://github.com/swri-robotics/descartes_light/issues/8